Remove default empty state of `sf::Texture`

ChrisThrasher opened the pull request:
## Description

Next in line after #3025 and others before it.

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -98,14 +91,12 @@ class SFML_GRAPHICS_API Texture : GlResource
     /// \param size Width and height of the texture
     ///
-    /// \return True if creation was successful
+    /// \return Texture if loading was successful, otherwise `std::nullopt`
```
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -119,17 +110,16 @@ class SFML_GRAPHICS_API Texture : GlResource
     /// \param area     Area of the image to load
     ///
-    /// \return True if loading was successful
+    /// \return Texture if loading was successful, otherwise `std::nullopt`
```
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

ChrisThrasher replied:
What makes you feel strongly about this order of words? There are lots of SFML APIs that use slightly different variations of this same sentence so there is no one standard phrasing for me to pick from. Because there is little consistency I don't put too much thought into it. Perhaps it would be worthwhile to go through all our docs and unify our phrasing for APIs with the same semantics.

vittorioromeo replied:
I don't feel strongly about it, it just sounds better in my mind. Feel free to ignore and I agree with your last sentence

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -143,18 +133,18 @@ class SFML_GRAPHICS_API Texture : GlResource
     /// \param area Area of the image to load
     ///
-    /// \return True if loading was successful
+    /// \return Texture if loading was successful, otherwise `std::nullopt`
```
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -119,17 +110,16 @@ class SFML_GRAPHICS_API Texture : GlResource
     ////////////////////////////////////////////////////////////
-    [[nodiscard]] bool loadFromFile(const std::filesystem::path& filename, const IntRect& area = IntRect());
+    [[nodiscard]] static std::optional<Texture> loadFromFile(const std::filesystem::path& filename,
+                                                             const IntRect&               area = IntRect());
```
```suggestion
                                                             const IntRect&               area = {});
```

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -143,18 +133,18 @@ class SFML_GRAPHICS_API Texture : GlResource
-    [[nodiscard]] bool loadFromMemory(const void* data, std::size_t size, const IntRect& area = IntRect());
+    [[nodiscard]] static std::optional<Texture> loadFromMemory(const void*    data,
+                                                               std::size_t    size,
+                                                               const IntRect& area = IntRect());
```
```suggestion
                                                               const IntRect& area = {});
```

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -168,17 +158,15 @@ class SFML_GRAPHICS_API Texture : GlResource
     /// \param area   Area of the image to load
     ///
-    /// \return True if loading was successful
+    /// \return Texture if loading was successful, otherwise `std::nullopt`
```
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -168,17 +158,15 @@ class SFML_GRAPHICS_API Texture : GlResource
     ///
     ////////////////////////////////////////////////////////////
-    [[nodiscard]] bool loadFromStream(InputStream& stream, const IntRect& area = IntRect());
+    [[nodiscard]] static std::optional<Texture> loadFromStream(InputStream& stream, const IntRect& area = IntRect());
```
```suggestion
    [[nodiscard]] static std::optional<Texture> loadFromStream(InputStream& stream, const IntRect& area = {});
```

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -192,17 +180,15 @@ class SFML_GRAPHICS_API Texture : GlResource
     /// \param area  Area of the image to load
     ///
-    /// \return True if loading was successful
+    /// \return Texture if loading was successful, otherwise `std::nullopt`
```
```suggestion
    /// \return Texture if loading was successful, `std::nullopt` otherwise
```

vittorioromeo commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -192,17 +180,15 @@ class SFML_GRAPHICS_API Texture : GlResource
     ///
     ////////////////////////////////////////////////////////////
-    [[nodiscard]] bool loadFromImage(const Image& image, const IntRect& area = IntRect());
+    [[nodiscard]] static std::optional<Texture> loadFromImage(const Image& image, const IntRect& area = IntRect());
```
```suggestion
    [[nodiscard]] static std::optional<Texture> loadFromImage(const Image& image, const IntRect& area = {});
```

vittorioromeo commented on src/SFML/Graphics/Texture.cpp:
```diff
@@ -175,21 +176,20 @@ bool Texture::create(const Vector2u& size)
 
     // Create the OpenGL texture if it doesn't exist yet
-    if (!m_texture)
+    if (!texture.m_texture)
```
```suggestion
```

This check is useless now, right?

ChrisThrasher replied:
Yeah I think you're right. Good catch.

vittorioromeo commented on src/SFML/Graphics/Texture.cpp:
```diff
@@ -198,7 +198,7 @@ bool Texture::create(const Vector2u& size)
                                          Context::isExtensionAvailable("GL_EXT_texture_edge_clamp");
 
-    if (!m_isRepeated && !textureEdgeClamp)
+    if (!texture.m_isRepeated && !textureEdgeClamp)
```
Ditto for the other checks -- unless I am missing something, they should all have their default value, as this is a brand new texture. Right?

ChrisThrasher replied:
`m_isRepeated` defaults to `false` and is never touched prior to this line so yeah that will always evaluate to `true` and thus can be removed.

I think this whole section can be removed as well.
```cpp
    if (texture.m_sRgb && !textureSrgb)
    {
        static bool warned = false;

        if (!warned)
        {
#ifndef SFML_OPENGL_ES
            err() << "OpenGL extension EXT_texture_sRGB unavailable" << '\n';
#else
            err() << "OpenGL ES extension EXT_sRGB unavailable" << '\n';
#endif
            err() << "Automatic sRGB to linear conversion disabled" << std::endl;

            warned = true;
        }

        texture.m_sRgb = false;
    }
    ```
`m_sRgb` defaults to `false` and is never touched prior to this line so I don't think there is ever a way to enter this conditional. This is a good sign because it shows that this API change systematically eliminates certain errors that were made possible by the previous API.

ChrisThrasher replied:
As I look closer, `Texture::setSrgb` is only useful if you call it prior to loading the texture. I think that needs to be added as a function parameter to ensure there is still a way to set this. I believe the same can be said for `Texture::setSmooth` and `Texture::setRepeated`.

ChrisThrasher commented on src/SFML/Graphics/Texture.cpp:
```diff
@@ -175,21 +176,20 @@ bool Texture::create(const Vector2u& size)
-    m_actualSize    = actualSize;
-    m_pixelsFlipped = false;
-    m_fboAttachment = false;
+    Texture texture;
+    texture.m_size       = size;
+    texture.m_actualSize = actualSize;
```
`sf::RenderTexture` takes advantage of this default constructor. To remove `Texture::Texture()` would result in more downstream changes to `RenderTexture`. However, if we also change `RenderTexture::create` to return a `std::optional<RenderTexture>` then this problem isn't so bad. I wonder if this PR should be postponed until I can perform this refactor inside `RenderTexture`. That order of operations may make more sense.

ChrisThrasher commented on src/SFML/Graphics/Font.cpp:
```diff
@@ -752,23 +752,28 @@ bool Font::setCurrentSize(unsigned int characterSize) const
-Font::Page::Page(bool smooth)
+Font::Page::Page(bool smooth) :
+texture(
+    [smooth]()
```
I thought this was a C++20 feature. I suppose I'll remove it and wait for CI to yell at me if it is in fact a C++20 feature.

ChrisThrasher commented:
I discovered a big issue with the current implementation. `sf::Texture::setSrgb`, `sf::Texture::setSmooth`, and `sf::Texture::setRepeated` are effectively used as constructor parameters and thus need to be reflected as parameters of the `create()` function. The definitions of those functions need to be inlined into `create()` as well so that functionality is not lost.

ChrisThrasher commented:
> I discovered a big issue with the current implementation. `sf::Texture::setSrgb`, `sf::Texture::setSmooth`, and `sf::Texture::setRepeated` are effectively used as constructor parameters and thus need to be reflected as parameters of the `create()` function.

I fixed this by adding parameters to all four static factory functions for setting the smooth, sRbg, and repeated properties. I believe this fixes the functionality issues. Now you can construct a texture with those settings set to something non-default.

The new problem is that an API with 3 consecutive `bool` parameters _sucks_. We need a better solution. I'm thinking it may be best to create a `sf::Texture::Settings` struct with 3 fields for these 3 parameters. Then we can use a default parameter to set this so that users who don't care don't have to write it out. See all the `false, false, false` I had to add hurts, but I'm sure a better solution is possible.

texus commented on include/SFML/Graphics/Texture.hpp:
```diff
@@ -663,9 +690,7 @@ SFML_GRAPHICS_API void swap(Texture& left, Texture& right) noexcept;
-/// sf::Texture texture;
-/// if (!texture.loadFromFile("texture.png"))
-///     return -1;
+/// const auto texture = sf::Texture::texture.loadFromFile("texture.png").value();
```
```suggestion
/// const auto texture = sf::Texture::loadFromFile("texture.png").value();
```

ChrisThrasher commented:
I need to take some more time with this PR so I'm marking it as a draft to indicate that it's not ready for review in its current state.

ChrisThrasher commented:
I took another shot at this PR. This time I removed `sf::Texture::setSrgb` since it seemingly had no more purpose as a member function. The only place that value is read is within the creation factory function. Setting that value after creation would do nothing so it seemed bug-prone to leave it there. Thus I added 1 parameter to the texture factory functions for specifying sRgb or not so this functionality is not lost.

ChrisThrasher commented on test/Graphics/Texture.test.cpp:
```diff
@@ -284,25 +252,20 @@ TEST_CASE("[Graphics] sf::Texture", runDisplayTests())
         CHECK_FALSE(texture1.isSrgb());
         CHECK(texture1.isSmooth());
         CHECK_FALSE(texture1.isRepeated());
-        CHECK(texture2.isSrgb());
```
Something to know about `setSrgb()` prior to this PR is that it must be called before `create()` for it to work. In this case it was called after `create()` meaning the texture didn't actually attempt to enable SRGB.

My API refactor means this parameter must be provided upon creation. As it turns out this feature doesn't work on Windows OpenGL ES so the `create()` function automatically disables it and emits a warning. That means we can't test for `isSrgb` being `true` since on that one config it's false.

kimci86 replied:
A comment explaining this in the code would be valuable so it doesn't look like testing it was just overlooked.

ChrisThrasher commented on src/SFML/Graphics/Font.cpp:
```diff
@@ -752,23 +752,28 @@ bool Font::setCurrentSize(unsigned int characterSize) const
+            for (unsigned int y = 0; y < 2; ++y)
+                image.setPixel({x, y}, Color::White);
+
+        // Create the texture
+        auto maybeTexture = sf::Texture::loadFromImage(image);
+        if (!maybeTexture)
+            err() << "Failed to load font page texture" << std::endl;
+
+        auto newTexture = maybeTexture.value();
+        newTexture.setSmooth(smooth);
+        return newTexture;
```
I kept this in my latest revision of this PR. I don't like what I wrote but can't picture anything better at the moment.

kimci86 replied:
The natural next step would be that Font::Page is created with a factory function.

ChrisThrasher replied:
I went ahead and made this change because it seems sensible. What happened is now the null check occurs downstream in `sf::Font::loadPage` where the pages are actually constructed. If construction fails inside `loadPage` then what do we do? I chose to `assert` that loading succeeds but we could just as well use `.value()` in which case we're back to the same solution as what I previously wrote. 

Should we go so far as to make `sf::Font::loadPage` return something nullable and put the onus users? That's possible but I don't love the idea.

ChrisThrasher commented:
Rebased on `master` and fixed merge conflicts.

vittorioromeo commented on examples/android/app/src/main/jni/main.cpp:
```diff
@@ -86,9 +86,7 @@ int main(int argc, char* argv[])
-    sf::Texture texture;
-    if (!texture.loadFromFile("image.png"))
-        return EXIT_FAILURE;
+    const auto texture = sf::Texture::loadFromFile("image.png").value();
```
This changes the behavior to throwing an exception rather than returning `EXIT_FAILURE`. Given that some behavior changes ended up being reverted in #3051, do we want to maintain the same behavior as before here too?

ChrisThrasher replied:
> This changes the behavior to throwing an exception

In practice the result is the same. Failure to load means the program stops immediately. 

> Given that some behavior changes ended up being reverted in https://github.com/SFML/SFML/pull/3051, do we want to maintain the same behavior as before here too?

#3051 was different. In that case I truly did change behavior in a way that totally broke the example on certain platforms. For now I want to use the same `.value()` pattern in our examples until we come to a consensus on a better alternative. I'm not sure exactly what alternative you or others may prefer.

vittorioromeo replied:
Makes sense, you are correct that #3051 was a special case. Still, would be nice to have a policy regarding exceptions in our examples. Resolving for now.

vittorioromeo commented on examples/cocoa/CocoaAppDelegate.mm:
```diff
@@ -57,7 +54,7 @@
     sf::Font              font{sf::Font::loadFromFile(resPath / "tuffy.ttf").value()};
     sf::Text              text{font};
-    sf::Texture           logo;
+    sf::Texture           logo{sf::Texture::loadFromFile(resPath / "logo.png").value()};
```
Ditto about behavioral changes, we also lose the `NSLog` call here. Note that we already changed behavior in line 55 for the font, so I'm not sure...

ChrisThrasher replied:
To be honest my attitude here was "meh whatever" about removing those logs. I'm not married to this solution and can go back to the previous solution if you feel strongly about keeping that.

vittorioromeo replied:
I never used/targeted Cocoa, I have no idea if those logs provide any real value or not, but honestly your solution is simpler and probably good enough. Unless you know anyone who targets Cocoa that can provide more feedback, I don't feel strongly about keeping the logs at all.

vittorioromeo commented on examples/opengl/OpenGL.cpp:
```diff
@@ -58,10 +58,7 @@ int main()
-        backgroundTexture.setSrgb(sRgb);
-        if (!backgroundTexture.loadFromFile(resourcesDir() / "background.jpg"))
-            return EXIT_FAILURE;
+        const auto       backgroundTexture = sf::Texture::loadFromFile(resourcesDir() / "background.jpg", sRgb).value();
```
Not going to repeat myself for all the other ones, but all of these `.value()` calls in the examples change behavior and such changes were reverted in #3051 IIRC.

Maybe we should have a policy about throwing in examples?

vittorioromeo commented on include/SFML/Graphics/Font.hpp:
```diff
@@ -315,7 +315,8 @@ class SFML_GRAPHICS_API Font
     {
-        explicit Page(bool smooth);
+        [[nodiscard]] static std::optional<Page> make(bool smooth);
+        Page(Texture&& texture);
```
```suggestion
        explicit Page(Texture&& texture);
```

vittorioromeo reviewed (changes requested):
Minor considerations but looks good to me otherwise, happy to approve once you take a look and decide what to do.

eXpl0it3r commented on src/SFML/Graphics/Font.cpp:
```diff
@@ -757,7 +760,7 @@ bool Font::setCurrentSize(unsigned int characterSize) const
 
 ////////////////////////////////////////////////////////////
-Font::Page::Page(bool smooth)
+std::optional<Font::Page> Font::Page::make(bool smooth)
```
Any reason we didn't use `create` as with most other factory functions?

ChrisThrasher replied:
[Good point](https://github.com/SFML/SFML/pull/3103)
